pkg/didcomm/common/messenger/messenger.go

```diff
@@ -0,0 +1,36 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package messenger
+
+import "github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
+
+// The messenger package is responsible for the handling of communication between agents.
+// It takes care of threading, timing, etc.
+// Each message that we are going to send should be DIDCommMsg type.
+// e.g we have type model.Ack{Type: "type", ID: "id"}
+// it should be converted to: map[@id:id @type:type]
+// NOTE: The package modifies message data by JSON tag name according to aries-rfcs.
+//       Fields like @id, ~thread , etc. are redundant and may be rewritten.
+
+// Client provides methods for the communication
+type Client interface {
+	// HandleMessage receives incoming message and prepares data needed for the communication according to the options.
+	// Options should be provided by the service.
+	HandleMessage(msg service.DIDCommMsg, myDID, theirDID string, opts ...interface{})
```

Comment:
What are the use cases for this method?

Reply:
HandleMessage will receive all incoming messages. According to `opts ...interface{}` which may be provided by any service, the behavior of this function may be different.

The function is responsible for saving some data which will be retrieved later by ReplyTo, Send, ReplyToNested functions. 

HandleMessage should provide correct sate for  ReplyTo, Send, ReplyToNested functions.
Actually, this function is responsible for counting incoming messages (probably there will be more logic), also it will provide us the possibility to get the threadID by msgID.

It is a wrapper function that will be executed by the dispatcher when we received an incoming message.

Reply:
I still don't understand.

The "messaging service" is a separate component from the protocol services. It will act as the "glue" between protocol services. Right?

Questions:

* What use is HandleMessage when this client can accept message via the other methods?
* How does a protocol service trigger another protocol using this interface?

Reply:
The `messenger` is a wrapper over dispatcher.  Instead of using `Outbound` you may use `messenger`. The `messenger` responsibility are:
* add message ID if it is needed
* take care of threading (thid,pthid,sender_order,received_orders)
* possibility to modify message payload(generic) automatically 

e.g You do not need to count sender_order or received_orders for the thread decorator.